pkg/cri/sbserver/podsandbox/sandbox_run.go

```diff
@@ -168,10 +181,10 @@ func (c *Controller) Start(ctx context.Context, id string) (_ *api.ControllerSta
 			volatileSandboxRootDir, err)
 	}
 	defer func() {
-		if retErr != nil {
+		if retErr != nil && cleanupErr == nil {
```

Comment:
If I understand this correctly, defers are called in reverse order, and if 1st cleanup func fails, we skip all remaining ones (for all subsequent ones `cleanupErr == nil` check will be failing). Like if we failed to `c.cleanupSandboxFiles`, we will not `c.os.RemoveAll(sandboxRootDir)`, right?
Should we instead use try calling all cleanup functions and aggregate errors happened along?


Something like:

```go
if retErr != nil {
   if err := cleanup(); err != nil {
       cleanUpErrs = append(cleanUpErrs, fmt.Errorf("failed to clean sandbox root dir: %w", err);
   }
}
```

Reply:
> If I understand this correctly, defers are called in reverse order, and if 1st cleanup func fails, we skip all remaining ones (for all subsequent ones cleanupErr == nil check will be false).

Correct, and this is intentional.  The problem addressed by https://github.com/containerd/containerd/pull/5904 relates to resource leaks created through partial cleanup; in other words: if one of the defers failed but the container representing the pod sandbox was removed, containerd would lose track of whichever resource had failed to be cleaned up.  (This is slightly different in sbserver since we have a persistent object representing the Sandbox, but the principle still applies; if that record is deleted then we lose track of resources.)  This was particularly troublesome for CNI cleanup failures, leading to things like future IPAM assignment failures (leaked IP addresses depleting the pool available).  The theory of operation for both #5904 and this port of it is that we halt cleanup at the first failure, retaining the record of allocated resources, and then the Kubelet will continue to reconcile and attempt to remove the pod sandbox at which point the resource cleanup operation is retried.